tests/xtro-sharpie/SimdCheck.cs

```diff
@@ -0,0 +1,382 @@
+				simd_type = typeName;
+			
+			return rv;
+		}
+
+		bool IsSimdType (QualType type, ref string simd_type, ref bool requires_marshal_directive)
+		{
+			var str = Undecorate (type.ToString ());
+
+			if (type_mapping.TryGetValue (str, out var info)) {
+				requires_marshal_directive = true;
+				simd_type = str;
+				return true;
+			}
+
+			if (IsExtVector (type, ref simd_type))
+				Console.WriteLine ($"!simd-unmapped-type! The Simd type {simd_type} does not have a mapping. Please add one in SimdCheck.cs");
+
+			return false;
+		}
+
+		string Undecorate (string native_name)
+		{
+			if (native_name.StartsWith ("const ", StringComparison.Ordinal))
+				return Undecorate (native_name.Substring ("const ".Length));
+
+			if (native_name.StartsWith ("struct ", StringComparison.Ordinal))
+				return Undecorate (native_name.Substring ("struct ".Length));
+
+			const string _Nonnull = " _Nonnull";
+			if (native_name.EndsWith (_Nonnull, StringComparison.Ordinal))
+				return Undecorate (native_name.Substring (0, native_name.Length - _Nonnull.Length));
+
+			const string _Nullable = " _Nullable";
+			if (native_name.EndsWith (_Nullable, StringComparison.Ordinal))
+				return Undecorate (native_name.Substring (0, native_name.Length - _Nullable.Length));
+			
+			const string _star = " *";
+			if (native_name.EndsWith (_star, StringComparison.Ordinal))
+				return Undecorate (native_name.Substring (0, native_name.Length - _star.Length));
+
+			return native_name;
+		}
+
+		bool HasMarshalDirective (ICustomAttributeProvider provider)
+		{
+			if (provider?.HasCustomAttributes != true)
+				return false;
+
+			foreach (var ca in provider.CustomAttributes)
+				if (ca.Constructor.DeclaringType.Name == "MarshalDirective")
+					return true;
+
+			return false;
+		}
+
+		void CheckMarshalDirective (MethodDefinition method, string simd_type)
+		{
+			if (!method.HasBody)
+				return;
+
+			// The [MarshalDirective] attribute isn't copied to the generated code,
+			// so instead apply some heuristics and detect calls to the xamarin_simd__ P/Invoke,
+			// and if there are any, then assume the method binding has a [MarshalDirective].
+			var body = method.Body;
+			var anyCalls = false;
+			foreach (var i in body.Instructions) {
+				switch (i.OpCode.Code) {
+				case Mono.Cecil.Cil.Code.Call:
+				case Mono.Cecil.Cil.Code.Calli:
+				case Mono.Cecil.Cil.Code.Callvirt:
+					var mr = i.Operand as MethodReference;
+					if (mr != null) {
+						if (mr.Name.StartsWith ("xamarin_simd__", StringComparison.Ordinal))
+							return;
+						if (mr.Name.StartsWith ("xamarin_vector_float3__", StringComparison.Ordinal))
+							return;
+					}
+					anyCalls = true;
+					break;
+				default:
+					break;
+				}
+			}
+
+			// If the method doesn't call anywhere, it can't be broken.
+			// For instance if the method just throws an exception.
+			if (!anyCalls)
+				return;
+
+			Console.WriteLine ($"!simd-missing-marshaldirective! {method}: simd type: {simd_type}");
```

Comment:
start with either `!unknown-` or `!wrong-` so it gets sorted along other important/critical logs